File: frontend/src/types/ui-template.ts
import type { BlockNode } from '@/ui-runtime/blocks'
import type { Device } from './experiment'

export type TemplateStatus = 'draft' | 'published'
export type TemplateType = 'tree' | 'code'

export interface UITemplate {
  id: string
  name: string
  description: string | null
  status: TemplateStatus
  template_type: TemplateType
  /** Target device for this template; one template covers exactly one device. */
  device: Device
  feed_config: Record<string, Record<string, string | number>>
  watch_config: Record<string, Record<string, string | number>>
  feed_css: string
  watch_css: string
  code_text: string | null
  /**
   * Phase 4 block-tree shape. Non-null when the template was authored
   * via the block-tree editor; null for legacy CSS-themed templates.
   * Renderer prefers `*_tree` over `*_config` when both are set.
   */
  feed_tree: BlockNode | null
  watch_tree: BlockNode | null
  created_at: string
  updated_at: string
}

export interface UITemplateListItem {
  id: string
  name: string
  description: string | null
  status: TemplateStatus
  template_type: TemplateType
  device: Device
  created_at: string
  updated_at: string
}

export interface UITemplateCreateRequest {
  name: string
  description?: string
  template_type?: TemplateType
  device?: Device
}

export interface UITemplateUpdateRequest {
  name?: string
  description?: string
  status?: TemplateStatus
  template_type?: TemplateType
  device?: Device
  feed_config?: Record<string, Record<string, string | number>>
  watch_config?: Record<string, Record<string, string | number>>
  feed_css?: string
  watch_css?: string
  code_text?: string
  feed_tree?: BlockNode | null
  watch_tree?: BlockNode | null
}

File: backend/app/models/ui_template.py
import uuid
from datetime import datetime
from sqlalchemy import Column, String, DateTime, Text
from sqlalchemy.dialects.postgresql import UUID, JSONB

from ..database import Base


class UITemplate(Base):
    __tablename__ = "ui_templates"

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    name = Column(String(255), nullable=False, unique=True)
    description = Column(String(1000), nullable=True)
    status = Column(String(20), nullable=False, default="draft")  # draft, published
    # 'tree' = visual block-tree authoring (feed_config / watch_config / *_css)
    # 'code' = raw TSX source authored in editor, compiled in browser at runtime
    template_type = Column(String(20), nullable=False, default="tree")
    feed_config = Column(JSONB, nullable=False, default=dict)
    watch_config = Column(JSONB, nullable=False, default=dict)
    feed_css = Column(Text, nullable=False, default="")
    watch_css = Column(Text, nullable=False, default="")
    code_text = Column(Text, nullable=True)
    # Phase 4 block-tree columns. When non-null and template_type='tree',
    # the renderer prefers these over the legacy feed_config/watch_config
    # shape. Stored as the BlockNode JSON shape declared in
    # frontend/src/ui-runtime/blocks/types.ts.
    feed_tree = Column(JSONB, nullable=True)
    watch_tree = Column(JSONB, nullable=True)
    created_at = Column(DateTime, default=datetime.utcnow)
    updated_at = Column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow)
